### PR TITLE
Week3 authenticate

### PR DESCRIPTION
**Pull request by KhawYewOnn**
Completed requirement for week 3

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -13,20 +13,80 @@
...
+
+    messages = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String message = (String) entity.getProperty("email");
+      String message2 = (String) entity.getProperty("message");
+      message = message + " said " + message2;
+      messages.add(message);
+    }
+
+    response.setContentType("application/json");
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:59]: Content type an be constant

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -13,20 +13,80 @@
...
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    messages = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String message = (String) entity.getProperty("email");
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:53]: Shouldn't variable name be email

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:18]: content type can be constant

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -13,20 +13,80 @@
...
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.*;
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:25]: Please avoid folded imports

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -13,20 +13,80 @@
...
 
+
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private ArrayList<String> messages;
+  
+  @Override
+  public void init() {
+    messages = new ArrayList<>();
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:41]: init can be with declaration

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -13,20 +13,80 @@
...
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    messages = new ArrayList<>();
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:51]: messages can just be a local variable it seems

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -13,20 +13,80 @@
...
+    for (Entity entity : results.asIterable()) {
+      String message = (String) entity.getProperty("email");
+      String message2 = (String) entity.getProperty("message");
+      message = message + " said " + message2;
+      messages.add(message);
+    }
+
+    response.setContentType("application/json");
+    String json = new Gson().toJson(messages);
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:60]: Gson does not need to be re-initialized very time this method is called and can just be a global declaration and initialization in this class

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -13,20 +13,80 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String message = getParameter(request, "text-input", "");
+    Entity commentEntity = new Entity("comment");
+    commentEntity.setProperty("message", message);
+    long timestamp = System.currentTimeMillis();
+    commentEntity.setProperty("timestamp", timestamp);
+    UserService userService = UserServiceFactory.getUserService();
+    // will fail if user is not logged in
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:73]: What is our fail safe if user is not logged in? Do we ask them to login?

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String userEmail = userService.getCurrentUser().getEmail();
+      String urlToRedirectToAfterUserLogsOut = "/index.html";
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+      result.add(userEmail);
+      result.add(logoutUrl);
+      /*
+      response.getWriter().println("<p>Hello " + userEmail + "!</p>");
+      response.getWriter().println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:29]: Remove stray comments

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+      */
+    } else {
+      String urlToRedirectToAfterUserLogsIn = "/index.html";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+      result.add(loginUrl);
+      /* 
+      response.getWriter().println("<p>Hello stranger.</p>");
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:36]: Remove stray comments

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+      response.getWriter().println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>"); 
+      */
+    }
+    String json = new Gson().toJson(result);
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:40]: Gson can be globally initialized in class.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -17,12 +17,78 @@
...
+ * Fetches stats from the servers and adds them to the DOM.
+ */
+function getComment() {
+  const statsListElement = document.getElementById('comment-container');
+  statsListElement.innerHTML = '';
+  var toStop = true;
+  var userEmail = "";
+  fetch('/login').then(response => response.json()).then((loginInfo) => {
+    /*var i;
+    for (i = 0; i < loginInfo.length; i++) {
```
Comment (rashikaranpuria) [inline portfolio/src/main/webapp/script.js:42]: remove stray comments

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -17,12 +17,78 @@
...
+      statsListElement.appendChild(createListElement(loginInfo[i]));
+    } */
+    if (loginInfo.length == 1) {
+      console.log("loginInfo.length == 1");
```
Comment (rashikaranpuria) [inline portfolio/src/main/webapp/script.js:46]: Can remove logging from production code, now

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -17,12 +17,78 @@
...
+      var parent = document.getElementById("content");
+      parent.appendChild(para);
+
+      fetch('/data').then(response => response.json()).then((stats) => {
+        // stats is an object, not a string, so we have to
+        // reference its fields to create HTML content
+
+        // stats should be empty if user is not logged in
+        var i;
+        for (i = 0; i < stats.length; i++) {
```
Comment (rashikaranpuria) [inline portfolio/src/main/webapp/script.js:80]: var i can be declared in for loop

File: portfolio/pom.xml [hunk 15]
```diff
@@ -1,6 +1,4 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
[the remark this answers is not in the document]
Comment (KhawYewOnn) [reply portfolio/pom.xml:1]: ok

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 16]
```diff
@@ -13,20 +13,80 @@
...
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.*;
+import java.util.ArrayList;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+
```
[the remark this answers is not in the document]
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: ok

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:59]: ok

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:53]: Oh yeah, I will change it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:51]: that's right, I will change it

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:73]: actually, this part can't be reached unless the user is logged in.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 9 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:29]: okay

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (KhawYewOnn) [reply portfolio/src/main/webapp/script.js:42]: okay

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (KhawYewOnn) [reply portfolio/src/main/webapp/script.js:46]: okay

Comment (KhawYewOnn) [reply portfolio/src/main/webapp/script.js:46]: okay

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (KhawYewOnn) [reply portfolio/src/main/webapp/script.js:80]: resolved

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:41]: ok

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (KhawYewOnn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:25]: fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 17]
```diff
@@ -13,20 +13,80 @@
...
+    }
+    return value;
+  }
+
+  private String convertEntityToMessage(Entity entity) {
+    return (String) entity.getProperty("message");
+  }
+
+  private String convertEntityToEmail(Entity entity) {
+    return (String) entity.getProperty("email");
```
Comment (rashikaranpuria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:90]: Abstracting these in separate methods is awesome

**Pull request merged**